### PR TITLE
feat(sdk-lib-mpc): add implementation of paillier-blum for paillier p…

### PR DESCRIPTION
**Pull request by zahin-mohammad**
- adds implementation for paillier blum proofs
- updates @types/node to match current node version (16.18.46)

TICKET: HSM-102, WP-132

### CODE REVIEW

File: modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+          .update(NBuf)
+          .update(wBuf)
+          .digest()
+      );
+      return h * h;
+    });
+}
+
+// https://en.wikipedia.org/wiki/Jacobi_symbol#Implementation_in_C++
+function jacobi(a, n): bigint {
```
Comment (zahin-mohammad) [inline modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts:33]: We can move this out to a generic math module in the future.

File: modules/sdk-lib-mpc/test/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (zahin-mohammad) [inline modules/sdk-lib-mpc/test/tsconfig.json:1]: Created this to allow tests import json files for fixtures.

File: modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    w = bigIntFromBufferBE(Buffer.from(await randBits(bitLength(N))));
+    if (jacobi(w, N) === BigInt(-1)) {
+      break;
+    }
+  }
+  // Prover generates y_i.
+  const y = generateY(N, w);
+  // Prover calculates z_i = y_i ^ d mod N
+  const z = y.map((y_i) => modPow(y_i, d, N));
+  // Prover calculates x_i = y_i ^ 1/4 mod N using [HOC - Fact 2.160]
```
Comment (islamaminBitGo) [inline modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts:92]: a link would be useful here, you can reference this https://cacr.uwaterloo.ca/hac/about/chap2.pdf

Comment (islamaminBitGo) [reply modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts:92]: you can change this to -> 
This is calculating the inverse of the function y^4 mod N, i.e. y^(1/4), where N = pq is a blum integer using HOC - Fact 2.160 from https://cacr.uwaterloo.ca/hac/about/chap2.pdf

File: modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+// Security parameter.
+const m = 80;
+
+/**
+ * Generate psuedo-random quadratic residue for (N, w, i).
+ * @param N - the prime number to verify is a product of two large primes.
+ * @param w - a random number with the same bitLength as N, that satisfies the Jacobi of w is -1 wrt N.
+ * @returns {bigint[]} - set of challenges for N
+ */
+function generateY(N, w): bigint[] {
```
Comment (islamaminBitGo) [inline modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts:16]: Im not sure where in this function does rejection sampling is done with the set $$Z^{*}_{N}$$ in mind. Need to do gcd(y,N) = 1 to check membership as described in Nothing-up-my-sleeve constructions spec here https://www.zkdocs.com/docs/zkdocs/protocol-primitives/nums/.

**Review by islamaminBitGo — CHANGES_REQUESTED** [review]

File: modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts (hunk 4 above)
Comment (zahin-mohammad) [reply modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts:16]: cc @johnoliverdriscoll

File: modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts (hunk 3 above)
Comment (zahin-mohammad) [reply modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts:92]:
```
  // This is calculating the inverse of the function y^4 mod N,
  // i.e.y ^ (1 / 4), where N = pq is a blum integer using HOC - Fact 2.160
  // from cacr.uwaterloo.ca / hac / about / chap2.pdf
  // Prover generates y_i.
  const y = generateY(N, w);
  // Prover calculates z_i = y_i ^ d mod N
  const z = y.map((y_i) => modPow(y_i, d, N));
```

like this?

File: modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  const NBuf = bigIntToBufferBE(N);
+  const wBuf = bigIntToBufferBE(w, NBuf.length);
+  return Array(m)
+    .fill(null)
+    .map((_, i) => {
+      const h = bigIntFromBufferBE(
+        createHmac('sha256', Buffer.from([i]))
```
Comment (islamaminBitGo) [inline modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts:23]: i think according to https://www.zkdocs.com/docs/zkdocs/protocol-primitives/nums/, we need to use an XOF hash function to produce a digest of the bitlength of N.

**Review by islamaminBitGo — CHANGES_REQUESTED** [review]

File: modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts (hunk 5 above)
Comment (islamaminBitGo) [reply modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts:23]: maybe something like SHAKE-256

Comment (Logicwax) [reply modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts:23]: sha3-XOF  (shake-256 mode) is typically used for that, especially when doing RSA keys

**Comment by zahin-mohammad** [comment]
Rebased off of master.

**Review by OttoAllmendinger — COMMENTED** [review]
flush (insufficient context)

**Comment by zahin-mohammad** [comment]
Rebased off of master

File: modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts (hunk 4 above)
Comment (johnoliverdriscoll) [reply modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts:16]: Done

File: modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts (hunk 5 above)
Comment (johnoliverdriscoll) [reply modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts]: Updated to use sha256 and separators for each field. However, TypeScript complains about the required parameters to createHash. See code comment.

**Review by OttoAllmendinger — CHANGES_REQUESTED** [review]
please summit the changes that are not part of the core change as a separate PR (e.g. https://github.com/BitGo/BitGoJS/pull/3806/commits/aab4f14685df1929dd316014f2b390ee0b19e770)

**Comment by zahin-mohammad** [comment]
> please summit the changes that are not part of the core change as a separate PR (e.g. [aab4f14](https://github.com/BitGo/BitGoJS/commit/aab4f14685df1929dd316014f2b390ee0b19e770))

Will do, moving that and a few other changes to seperate prs. This will stay in draft until those are merged.

**Review by sachushaji — COMMENTED** [review]
Flushing, Kindly re-request when ready

**Comment by zahin-mohammad** [comment]
Closing in place of https://github.com/BitGo/BitGoJS/pull/3940

**Pull request closed without merging**